File: test/python/steps/eligibility_dashboard_steps.py
# -*- coding: utf-8 -*-
"""
Course Eligibility Dashboard Step Definitions

This module implements BDD steps for testing the eligibility dashboard that shows
users which courses/forms they are eligible for based on their completed prerequisites.

This is NEW functionality not present in the legacy codebase, implemented as a test-side mock.
"""

from behave import given, when, then


# Directly include the helper functions here to avoid import-time side effects
# from form_prerequisites_steps
def init_prerequisites_context(context):
    """Initialize the prerequisites context for form availability testing."""
    if not hasattr(context, 'course_completions'):
        context.course_completions = {}

    if not hasattr(context, 'available_forms'):
        context.available_forms = []


def update_available_forms(context):
    """Update available forms based on course completions."""
    completions = context.course_completions

    available = []

    # TTC Application always available
    available.append('ttc_application')

    # DSN requires Happiness Program
    if completions.get('happiness_program'):
        available.append('dsn_application')

    # Part 1 requires Happiness Program
    if completions.get('happiness_program'):
        available.append('part_1_application')

    # Part 2 requires Part 1
    if completions.get('part_1'):
        available.append('part_2_application')

    # YES+ requires Part 1 and Part 2
    if completions.get('part_1') and completions.get('part_2'):
        available.append('yes_plus_application')

    context.available_forms = available


def get_eligibility_courses(context):
    """Return list of courses with eligibility status based on course completions."""
    # Initialize prerequisites context if needed
    init_prerequisites_context(context)

    completions = getattr(context, 'course_completions', {
        'happiness_program': False,
        'part_1': False,
        'part_2': False
    })

    # For TTC Evaluation, check if TTC was submitted
    ttc_submitted = completions.get('ttc_submitted', False)

    courses = [
        {
            'course': 'TTC Application',
            'prerequisite': 'None',
            'status': 'Eligible'
        },
        {
            'course': 'TTC Evaluation',
            'prerequisite': 'TTC Application submitted',
            'status': 'Eligible' if ttc_submitted else 'Not Eligible'
        },
        {
            'course': 'DSN Application',
            'prerequisite': 'Happiness Program completed',
            'status': 'Eligible' if completions.get('happiness_program') else 'Not Eligible'
        },
        {
            'course': 'Part 1',
            'prerequisite': 'Happiness Program completed',
            'status': 'Eligible' if completions.get('happiness_program') else 'Not Eligible'
        },
        {
            'course': 'Part 2',
            'prerequisite': 'Part 1 completed',
            'status': 'Eligible' if completions.get('part_1') else 'Not Eligible'
        }
    ]

    return courses


def init_eligibility_dashboard_context(context):
    """Initialize the eligibility dashboard context."""
    # Initialize prerequisites context
    init_prerequisites_context(context)
    update_available_forms(context)

    # Initialize eligibility dashboard context
    if not hasattr(context, 'eligibility_dashboard'):
        context.eligibility_dashboard = {
            'courses': [],
            'form_messages': {}
        }

    if not hasattr(context, 'form_access_attempt'):
        context.form_access_attempt = None


# Step: I view my course eligibility dashboard
@when('I view my course eligibility dashboard')
def step_view_eligibility_dashboard(context):
    """Initialize and populate the eligibility dashboard with course list."""
    init_eligibility_dashboard_context(context)
    context.eligibility_dashboard['courses'] = get_eligibility_courses(context)


# Step: I should see a list of available courses with prerequisites:
# Note: Both with and without colon to handle behave version differences
@then('I should see a list of available courses with prerequisites:')
@then('I should see a list of available courses with prerequisites')
def step_see_course_list(context):
    """Verify the eligibility dashboard shows the expected courses with prerequisites and status."""
    # Get expected courses from the data table
    expected = []
    for row in context.table:
        expected.append({
            'course': row['course'],
            'prerequisite': row['prerequisite'],
            'status': row['status']
        })

    # Get actual courses from dashboard
    actual = context.eligibility_dashboard.get('courses', [])

    # Verify all expected courses are present
    for exp_course in expected:
        found = False
        for act_course in actual:
            if (act_course['course'] == exp_course['course'] and
                act_course['prerequisite'] == exp_course['prerequisite'] and
                act_course['status'] == exp_course['status']):
                found = True
                break

        assert found, \
            'Expected course not found in dashboard: %s (got: %s)' % (exp_course, actual)


# Step: I attempt to access the DSN application form
@when('I attempt to access the DSN application form')
def step_attempt_access_dsn_form(context):
    """Record attempt to access DSN form and determine eligibility."""
    init_eligibility_dashboard_context(context)

    # Check if DSN is available based on prerequisites
    is_available = 'dsn_application' in context.available_forms

    context.form_access_attempt = 'dsn_application'

    # Store the message for this form
    if is_available:
        context.eligibility_dashboard['form_messages']['dsn_application'] = {
            'available': True,
            'message': 'available',
            'explanation': None
        }
    else:
        context.eligibility_dashboard['form_messages']['dsn_application'] = {
            'available': False,
            'message': 'not available',
            'explanation': 'Complete Happiness Program first'
        }


# Step: I should see "not available" message
@then('I should see "not available" message')
def step_see_not_available_message(context):
    """Verify that a not available message is shown."""
    assert context.form_access_attempt is not None, \
        'No form access attempt recorded'

    form = context.form_access_attempt
    message_data = context.eligibility_dashboard.get('form_messages', {}).get(form, {})

    assert message_data.get('message') == 'not available', \
        'Expected "not available" message, got: %s' % message_data


# Step: the message should explain the prerequisite: "{string}"
@then('the message should explain the prerequisite: "{explanation}"')
def step_explain_prerequisite(context, explanation):
    """Verify that the prerequisite explanation matches expected."""
    assert context.form_access_attempt is not None, \
        'No form access attempt recorded'

    form = context.form_access_attempt
    message_data = context.eligibility_dashboard.get('form_messages', {}).get(form, {})

    actual_explanation = message_data.get('explanation')
    assert actual_explanation == explanation, \
        'Expected explanation "%s", got: "%s"' % (explanation, actual_explanation)


# Step: the DSN form shows as "not available"
@given('the DSN form shows as "not available"')
@then('the DSN form shows as "not available"')
def step_dsn_shows_not_available(context):
    """Verify that DSN form is marked as not available."""
    init_prerequisites_context(context)
    update_available_forms(context)

    assert 'dsn_application' not in context.available_forms, \
        'DSN application should not be available, got: %s' % context.available_forms


# Step: I refresh the eligibility dashboard
@when('I refresh the eligibility dashboard')
def step_refresh_eligibility_dashboard(context):
    """Re-calculate eligibility based on current course completions."""
    init_eligibility_dashboard_context(context)

    # Re-calculate courses with updated eligibility
    context.eligibility_dashboard['courses'] = get_eligibility_courses(context)


# Step: the DSN form should show as "available"
@then('the DSN form should show as "available"')
def step_dsn_shows_available(context):
    """Verify that DSN form is marked as available."""
    init_prerequisites_context(context)
    update_available_forms(context)

    assert 'dsn_application' in context.available_forms, \
        'DSN application should be available, got: %s' % context.available_forms

File: test/typescript/steps/eligibility_dashboard_steps.ts
/**
 * Course Eligibility Dashboard Step Definitions
 *
 * This module implements BDD steps for testing the eligibility dashboard that shows
 * users which courses/forms they are eligible for based on their completed prerequisites.
 *
 * This is NEW functionality not present in the legacy codebase, implemented as a test-side mock.
 */

import { DataTable, Given, When, Then } from '@cucumber/cucumber';

// Import prerequisites context from form_prerequisites_steps
import {
  prerequisitesContext,
  initPrerequisitesContext,
  updateAvailableForms,
} from './form_prerequisites_steps';

/**
 * Course eligibility interface
 */
interface Course {
  course: string;
  prerequisite: string;
  status: string;
}

/**
 * Form message interface
 */
interface FormMessage {
  available: boolean;
  message: string;
  explanation: string | null;
}

/**
 * Eligibility dashboard context
 */
interface EligibilityDashboard {
  courses: Course[];
  formMessages: Record<string, FormMessage>;
}

export const eligibilityDashboardContext: EligibilityDashboard = {
  courses: [],
  formMessages: {},
};

export let formAccessAttempt: string | null = null;

/**
 * Get list of courses with eligibility status based on course completions.
 */
function getEligibilityCourses(): Course[] {
  const completions = prerequisitesContext.course_completions;
  const ttc_submitted = completions.ttc_submitted || false;

  return [
    {
      course: 'TTC Application',
      prerequisite: 'None',
      status: 'Eligible',
    },
    {
      course: 'TTC Evaluation',
      prerequisite: 'TTC Application submitted',
      status: ttc_submitted ? 'Eligible' : 'Not Eligible',
    },
    {
      course: 'DSN Application',
      prerequisite: 'Happiness Program completed',
      status: completions.happiness_program ? 'Eligible' : 'Not Eligible',
    },
    {
      course: 'Part 1',
      prerequisite: 'Happiness Program completed',
      status: completions.happiness_program ? 'Eligible' : 'Not Eligible',
    },
    {
      course: 'Part 2',
      prerequisite: 'Part 1 completed',
      status: completions.part_1 ? 'Eligible' : 'Not Eligible',
    },
  ];
}

/**
 * Initialize the eligibility dashboard context.
 */
function initEligibilityDashboardContext(): void {
  initPrerequisitesContext();
  updateAvailableForms();

  eligibilityDashboardContext.courses = [];
  eligibilityDashboardContext.formMessages = {};
  formAccessAttempt = null;
}

/**
 * Reset eligibility dashboard state between test scenarios.
 * Called by common.ts Before hook to prevent state leakage.
 */
export function resetEligibilityDashboardState(): void {
  eligibilityDashboardContext.courses = [];
  eligibilityDashboardContext.formMessages = {};
  formAccessAttempt = null;
}

// Step: I view my course eligibility dashboard
When('I view my course eligibility dashboard', function () {
  initEligibilityDashboardContext();
  eligibilityDashboardContext.courses = getEligibilityCourses();
});

// Step: I should see a list of available courses with prerequisites:
// Note: Both with and without colon to handle cucumber version differences
Then('I should see a list of available courses with prerequisites:', function (dataTable: DataTable) {
  expectCoursesToMatch(dataTable);
});

Then('I should see a list of available courses with prerequisites', function (dataTable: DataTable) {
  expectCoursesToMatch(dataTable);
});

/**
 * Helper to verify courses match expected data table
 */
function expectCoursesToMatch(dataTable: DataTable): void {
  const rawTable = dataTable.raw();
  if (rawTable.length === 0) {
    throw new Error('Expected eligibility table to include a header row');
  }

  const headerRow = rawTable[0].map((column) => column.trim());
  const normalizedHeader = headerRow.map((column) => column.toLowerCase());
  const requiredColumns = ['course', 'prerequisite', 'status'] as const;
  const columnIndex: Record<(typeof requiredColumns)[number], number> = {
    course: -1,
    prerequisite: -1,
    status: -1,
  };

  for (const [index, column] of normalizedHeader.entries()) {
    if (requiredColumns.includes(column as (typeof requiredColumns)[number])) {
      columnIndex[column as (typeof requiredColumns)[number]] = index;
    }
  }

  const missingColumns = requiredColumns.filter((column) => columnIndex[column] === -1);
  if (missingColumns.length > 0) {
    throw new Error(
      `Eligibility table missing required columns: ${missingColumns.join(', ')} (got: ${headerRow.join(', ')})`,
    );
  }

  if (rawTable.length <= 1) {
    throw new Error('Expected eligibility table to include at least one course row');
  }

  const expected: Course[] = rawTable.slice(1).map((row, index) => {
    const course = row[columnIndex.course];
    const prerequisite = row[columnIndex.prerequisite];
    const status = row[columnIndex.status];

    if (!course || !prerequisite || !status) {
      throw new Error(
        `Eligibility table row ${index + 1} must include course, prerequisite, and status`,
      );
    }

    return {
      course,
      prerequisite,
      status,
    };
  });

  if (expected.length === 0) {
    throw new Error('Expected eligibility table to include at least one course row');
  }

  const actual = eligibilityDashboardContext.courses;
  const actualByCourse = new Map<string, Course>(
    actual.map((course) => [course.course, course]),
  );

  // Verify all expected courses are present with matching prerequisite and status.
  for (const expCourse of expected) {
    const actualCourse = actualByCourse.get(expCourse.course);
    if (!actualCourse) {
      throw new Error(
        `Expected course not found in dashboard: ${expCourse.course} (got: ${JSON.stringify(actual)})`,
      );
    }

    if (
      actualCourse.prerequisite !== expCourse.prerequisite ||
      actualCourse.status !== expCourse.status
    ) {
      throw new Error(
        `Course "${expCourse.course}" mismatch. Expected prerequisite "${expCourse.prerequisite}" and status "${expCourse.status}", ` +
          `got prerequisite "${actualCourse.prerequisite}" and status "${actualCourse.status}".`,
      );
    }
  }
}

// Step: I attempt to access the DSN application form
When('I attempt to access the DSN application form', function () {
  initEligibilityDashboardContext();

  const forms = prerequisitesContext.available_forms || [];
  const isAvailable = forms.includes('dsn_application');

  formAccessAttempt = 'dsn_application';

  if (isAvailable) {
    eligibilityDashboardContext.formMessages['dsn_application'] = {
      available: true,
      message: 'available',
      explanation: null,
    };
  } else {
    eligibilityDashboardContext.formMessages['dsn_application'] = {
      available: false,
      message: 'not available',
      explanation: 'Complete Happiness Program first',
    };
  }
});

// Step: I should see "not available" message
Then('I should see "not available" message', function () {
  if (formAccessAttempt === null) {
    throw new Error('No form access attempt recorded');
  }

  const form = formAccessAttempt;
  const messageData = eligibilityDashboardContext.formMessages[form];

  if (!messageData) {
    throw new Error(`No message data found for form: ${form}`);
  }

  if (messageData.message !== 'not available') {
    throw new Error(
      `Expected "not available" message, got: ${JSON.stringify(messageData)}`,
    );
  }
});

// Step: the message should explain the prerequisite: "{string}"
// Note: Both patterns for different cucumber versions
Then('the message should explain the prerequisite: "{string}"', function (explanation: string) {
  assertExplanationMatches(explanation);
});

Then('the message should explain the prerequisite: {string}', function (explanation: string) {
  assertExplanationMatches(explanation);
});

/**
 * Helper to verify explanation matches
 */
function assertExplanationMatches(explanation: string): void {
  if (formAccessAttempt === null) {
    throw new Error('No form access attempt recorded');
  }

  const form = formAccessAttempt;
  const messageData = eligibilityDashboardContext.formMessages[form];

  if (!messageData) {
    throw new Error(`No message data found for form: ${form}`);
  }

  if (messageData.explanation !== explanation) {
    throw new Error(
      `Expected explanation "${explanation}", got: "${messageData.explanation}"`,
    );
  }
}

// Step: the DSN form shows as "not available"
// Note: Used as Given in the feature file
Given('the DSN form shows as "not available"', function () {
  initPrerequisitesContext();
  updateAvailableForms();

  const forms = prerequisitesContext.available_forms || [];
  if (forms.includes('dsn_application')) {
    throw new Error(
      `DSN application should not be available, got: ${JSON.stringify(forms)}`,
    );
  }
});

// Step: I refresh the eligibility dashboard
When('I refresh the eligibility dashboard', function () {
  initEligibilityDashboardContext();
  eligibilityDashboardContext.courses = getEligibilityCourses();
});

// Step: the DSN form should show as "available"
Then('the DSN form should show as "available"', function () {
  initPrerequisitesContext();
  updateAvailableForms();

  const forms = prerequisitesContext.available_forms || [];
  if (!forms.includes('dsn_application')) {
    throw new Error(
      `DSN application should be available, got: ${JSON.stringify(forms)}`,
    );
  }
});
